### DISCUSSION TITLE
Welcome to Bash-Oneliner Discussions!

### DISCUSSION THREAD

**Discussion by onceupon** [onceupon/Bash-Oneliner#17] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Bushmills** [comment]
As there's no section dealing with brace expansion, here's a useful example which can reduce typos and save keystrokes:

`mv long_and_complicated_filename.{foo,bar}`

This, when executed, expands to:

`mv long_and_complicated_filename.foo long_and_complicated_filename.bar`

**Reply by mu578** [reply]
In that case it is named `filename expansion`. I think the default mode today is ERE (Extended Regular Expression, to handle more complex glob patterns, space, unicodes etc).

see:
- glob patterns
- brace expansion
- filename expansion
- learn awk

```
touch filename[0-9]   // existing files
touch filename{0..9} // create files
```

**Comment by Bushmills** [comment]
Frequently one wished to place nicely formatted headers into output, consisting of a line of specific chars, then a line of output, and another line of specific chars, those with the same length as the output line. That can be done like this:

`line="foo bar"; printf "%s\n%s\n%s\n" "${line//?/=}" "$line" "${line//?/=}"`

"line" may of course be passed as argument, when writing the above code as function. Also, header can be customised, for example with leading/trailing "#" chars according personal taste and need. Idea here is merely to generate header bracketing lines with the same length as the header line itself, by replacing all of its characters against the desired header character.

With leading "#" chars the line could look like this:
`line="foo bar"; printf "# %s\n# %s\n# %s\n" "${line//?/=}" "$line" "${line//?/=}"`
